pkg/ring/ring.go

```diff
@@ -255,57 +255,76 @@ func NewWithStoreClientAndStrategy(cfg Config, name, key string, store kv.Client
 		),
 	}
 
-	r.Service = services.NewBasicService(nil, r.loop, nil).WithName(fmt.Sprintf("%s ring client", name))
+	r.Service = services.NewBasicService(r.starting, r.loop, nil).WithName(fmt.Sprintf("%s ring client", name))
 	return r, nil
 }
 
+func (r *Ring) starting(ctx context.Context) error {
+	// Get the initial ring state so that, as soon as the service will be running, the in-memory
+	// ring would be already populated and there's no race condition between when the service is
+	// running and the WatchKey() callback is called for the first time.
+	value, err := r.KVClient.Get(ctx, r.key)
+	if err != nil {
+		return errors.Wrap(err, "unable to initialise ring state")
+	}
+	if value == nil {
+		level.Info(log.Logger).Log("msg", "ring doesn't exist in KV store yet")
+		return nil
+	}
+
+	r.updateRingState(value.(*Desc))
+	return nil
+}
+
 func (r *Ring) loop(ctx context.Context) error {
 	r.KVClient.WatchKey(ctx, r.key, func(value interface{}) bool {
 		if value == nil {
 			level.Info(log.Logger).Log("msg", "ring doesn't exist in KV store yet")
 			return true
 		}
 
-		ringDesc := value.(*Desc)
-
-		r.mtx.RLock()
-		prevRing := r.ringDesc
-		r.mtx.RUnlock()
-
-		rc := prevRing.RingCompare(ringDesc)
-		if rc == Equal || rc == EqualButStatesAndTimestamps {
-			// No need to update tokens or zones. Only states and timestamps
-			// have changed. (If Equal, nothing has changed, but that doesn't happen
-			// when watching the ring for updates).
-			r.mtx.Lock()
-			r.ringDesc = ringDesc
-			r.mtx.Unlock()
-			return true
-		}
-
-		now := time.Now()
-		ringTokens := ringDesc.GetTokens()
-		ringTokensByZone := ringDesc.getTokensByZone()
-		ringInstanceByToken := ringDesc.getTokensInfo()
-		ringZones := getZones(ringTokensByZone)
-
-		r.mtx.Lock()
-		defer r.mtx.Unlock()
-		r.ringDesc = ringDesc
-		r.ringTokens = ringTokens
-		r.ringTokensByZone = ringTokensByZone
-		r.ringInstanceByToken = ringInstanceByToken
-		r.ringZones = ringZones
-		r.lastTopologyChange = now
-		if r.shuffledSubringCache != nil {
-			// Invalidate all cached subrings.
-			r.shuffledSubringCache = make(map[subringCacheKey]*Ring)
-		}
+		r.updateRingState(value.(*Desc))
 		return true
 	})
 	return nil
 }
 
+func (r *Ring) updateRingState(ringDesc *Desc) {
```

Comment:
This function just contains what previously was in the `WatchKey()` callback. No logic changes.